### ISSUE TITLE
Performance monitoring

### ISSUE THREAD

**Issue by vikobeibe** [gdnsd/gdnsd#182]
Hello,
first of all thank you for your awesome contribution. We are using GDNSD in production and it has been extremely stable and we love it. 
Is there a way to log the time it takes to complete each request? Or any other way to debug what may cause two identical requests, issued one ofter another, from the same location via the same network, to sometimes have a difference in completion time up to 900ms? We've (mostly) ruled out network issues at this point.

Thanks in advance!

**Comment by blblack** [comment]
It probably doesn't make a ton of sense to instrument for local response times within gdnsd.  In theory we could put some perf hooks around measuring time from packet reception to packet emission (in other words, time from `recvmsg()` to `sendmsg()`), but:

1. The performance measurement and logging itself would probably be a performance reducer on its own.  The design of gdnsd is such that for the high-traffic UDP case, in that span between `recvmsg()` and `sendmsg()`, there is no dynamic memory allocation and no other syscalls.  It's all just memory copying between existing allocations and network buffers and code/logic, and none of the logic is so heavy as to account for 900ms of execution with any reasonable configuration on any reasonable system.  Getting high-resolution wallclock timestamps would probably be a significant burden, relatively-speaking.  Linux does, however, offer coarse-grained clocks like `CLOCK_MONOTONIC_COARSE` which wouldn't present much perf burden, which I believe on typical x86_64 kernels would give ~10ms resolution; not enough to really measure gdnsd's code performance, but easily enough to detect some kind of 900ms stall during a single request->response cycle.  It wouldn't be hard to patch in a quick hack using that and e.g. printing a log message when >=100ms is detected to find the outlier cases.

2. Even if we had such a thing, it probably would fail to provide useful results.  Either it would show evidence of a 900ms delay in responding, or it wouldn't.  If it did show such a thing, then there are two possible explanations: a serious design bug in gdnsd's runtime that's causing some kind of stall (it wouldn't be something simple that could be config-tuned away I don't think), or it's just going to show you that something else outside of gdnsd's code (the kernel and the rest of the processes on the system, etc) somehow pre-empted gdnsd being able to keep up (which is already a good guess, and the measurement here won't help you find it).  Either that or it doesn't show the problematic high latency values at all, but there could still be a system-local source of delay that's outside the boundary of gdnsd's receiving and sending syscalls (something off in the kernel, network stack, output queues, etc).

900ms is also a pretty crazy value in general for it to just be a "performance" problem.  It's orders of magnitude beyond what one would expect from any kind of performance tuning problem for a fast DNS server.  If we were talking about sub-ms variations I might ask about thread/socket counts, NUMA domains for the card and the daemon, socket buffer sizes and various sysctl tweaks, etc... but almost none of the normal perf-tuning things could believably make things that slow on a reasonable system.

If you want to bisect the problem space of "where did this 900ms happen at?" a bit, I would suggest:

1. Look for retransmits at the client side by capturing DNS packets with wireshark/tcpdump.  It's possible the initial query packet or the response to the initial query packet are being dropped somewhere (in the network, or by either side's network stack due to some queue or buffer overflowing at some layer), and the ~900ms represents the time for the client-side implementation to timeout waiting on a response and try again one or more times before finally getting a roundtrip response without drops.  There could be other evidence of various types of host-level drops as well (interface statistics), or gdnsd's own `dropped`, `sendfail`, and `recvfail` stats.

2. Look at the overall system performance of the client and the server under test.  Is something else going on with load and/or network traffic on these boxes outside of the DNS stuff you're looking at, which could potentially significantly interference with client or server execution or network traffic?  Periodic heavy load from something related to config management or monitoring for instance, or other production daemons sharing the machine?

3. If you're not already doing so, set the gdnsd configuration option `lock_mem = true` (which will require appropriate locked-memory ulimits, or infinity) to avoid the potential problem of any of gdnsd's memory getting swapped out under system memory pressure.  See also the priority tuning stuff in the default systemd unit file (which sets the daemon to a nice value of `-11`).

If we want to dig into the gdnsd side more, I'd probably need some details on gdnsd's config and version, the host platform/environment, etc.  Especially if the slow responses are dynamic plugin responses as opposed to static zonefile data.  Tracing out a server-side problem via `perf` and similar tools (assuming modern Linux) might help as well.

**Comment by vikobeibe** [comment]
Hello and thank you for the detailed response. 
I have since identified that I am likely wrong in excluding the network as a factor. After some more monitoring, I see that the domains registered with one particular registrar are so far the only ones to have this issue. I have also managed to setup another machine in the same DC where one of the GDNS servers is and interact with it over a VLAN, and it is responding very fast with no delays at all, and the performance does not vary by more than 1-2ms. As such I think I may have wasted your time for which I am sorry. I don't think this is at all a GDNSD issue. Thank you once again for your great work!
